### PR TITLE
Support for input pins.

### PR DESCRIPTION
**Pull request by Atmelfan**
Added support for embedded-hal InputPin.

I only made a basic proof-of-concept by modifying `examples/blink.rs` and the API is a bit awkward so feedback would be appreciated.

Tested with a FT2232.

### CODE REVIEW

**Comment by Atmelfan** [comment]
I would also want to change the current behavior of setting all GPIOs to outputs to setting them as inputs by default, to avoid any GPIOs from being damaged when connected to another output. Then setting the direction when calling `adx()/adix()`.

Might break a lot existing code...

**Comment by newAM** [comment]
Thanks for the pull request!  I am pretty busy right now, I will review this next weekend at the latest.

**Review by newAM — APPROVED** [review]
Code changes look good to me.  Thanks for adding this functionality!  I added two minor commits, let me know if those changes are ok!

I do not have access to FTDI hardware right now, I will merge this and make a release after I can sanity check it with some physical hardware.

**Comment by newAM** [comment]
> I would also want to change the current behavior of setting all GPIOs to outputs to setting them as inputs by default, to avoid any GPIOs from being damaged when connected to another output. Then setting the direction when calling `adx()/adix()`.
> 
> Might break a lot existing code...

This sounds like a good change to make, I would be ok with it.

**Comment by geomatsi** [comment]
Hi @Atmelfan 

Thanks for your PR !

I would suggest to add another simple 'button' example instead of changing 'blink' example. Besides, do you mind to cleanup history a little bit by rebasing and squashing fixups into appropriate commits ?

As for the API, originally I thought about adding something similar to what we have in various stm32 HALs:
```
let mut led = hal.gpio().ad3().into_output();
let mut button = hal.gpio().ad5().into_input();
```
But it looks like it does not add much value here other than writing more rust code inside this crate )
So lets keep it simple and start using inputs with your approach. I will also help with testing on different chips during the upcoming weekend.

Regards,
Sergey

File: src/gpio.rs [hunk 1]
```diff
@@ -73,3 +81,79 @@ where
         self.set(true)
     }
 }
+
+/// FTDI input pin.
+///
+/// This is created by calling [`FtHal::ad0`] - [`FtHal::ad7`].
```
Comment (geomatsi) [inline src/gpio.rs:87]: ad -> adi

File: src/gpio.rs [hunk 2]
```diff
@@ -73,3 +81,79 @@ where
...
+///
+/// [`FtHal::ad0`]: crate::FtHal::ad0
```
Comment (geomatsi) [inline src/gpio.rs:89]: ad0 -> adi0

File: src/gpio.rs [hunk 3]
```diff
@@ -73,3 +81,79 @@ where
...
+/// [`FtHal::ad7`]: crate::FtHal::ad7
```
Comment (geomatsi) [inline src/gpio.rs:90]: ad7 -> adi7

**Comment by geomatsi** [comment]
> I would also want to change the current behavior of setting all GPIOs to outputs to setting them as inputs by default, to avoid any GPIOs from being damaged when connected to another output. Then setting the direction when calling `adx()/adix()`.

IIRC for this purpose you will have to change default direction value (0xFB) in FtInner From implementation, see lib.rs.

**Comment by geomatsi** [comment]
Tested on both ft232h and ft2232h. Looks good !

From my experiments, it looks like input pins are pulled up by default. Could find anything about that in FTDI docs so far...

**Comment by Atmelfan** [comment]
Hi, sorry for not responding earlier. 

I've hit some weird corner case when doing a large SPI transaction (~100kB in 1kB parts) followed by a pin read. The GPIO read returns no data bytes and panics. 

Inserting a delay of a few milliseconds between the transfer and GPIO read solves the issue.

I'll try to make an example to reproduce it later today.

**Comment by Atmelfan** [comment]
An example which writes 104kB before a gpio read:

```
// examples/crash.rs
use embedded_hal::blocking::spi::Write;
use embedded_hal::digital::v2::{InputPin, OutputPin};
use ftdi_embedded_hal as hal;
use std::{thread::sleep, time::Duration};

const NUM_BLINK: usize = 10;
const SLEEP_DURATION: Duration = Duration::from_millis(500);

#[cfg(all(feature = "ftdi", feature = "libftd2xx"))]
compile_error!("features 'ftdi' and 'libftd2xx' cannot be enabled at the same time");

#[cfg(not(any(feature = "ftdi", feature = "libftd2xx")))]
compile_error!("one of features 'ftdi' and 'libftd2xx' shall be enabled");

fn main() {
    #[cfg(feature = "libftd2xx")]
    let device: libftd2xx::Ft232h = libftd2xx::Ftdi::new().unwrap().try_into().unwrap();

    #[cfg(feature = "ftdi")]
    let device = ftdi::find_by_vid_pid(0x0403, 0x6010)
        .interface(ftdi::Interface::A)
        .open()
        .unwrap();

    let hal = hal::FtHal::init_freq(device, 3000000).expect("Failed to init device");
    let mut spi = hal.spi().unwrap();
    let mut done = hal.adi6().unwrap();

    let bitstream = [0u8; 104090];

    // Send configuration image serially
    for chunk in bitstream.chunks(1024) {
        spi.write(chunk).unwrap();
    }

    // CDONE = 1?
    let cdone = done.is_high().unwrap();
    if !cdone {
        // Error
        println!("Not done?");
    }
}

```

Panics with:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Error { kind: UnexpectedEof, message: "failed to fill whole buffer" })'
```

The error seems to originate here (in gpio.rs):

```
let cmd: MpsseCmdBuilder = MpsseCmdBuilder::new().gpio_lower().send_immediate();
inner.ft.send(cmd.as_slice())?;
inner.ft.recv(&mut buffer)?;
```
It seems to work with `libftd2xx` but not `ftdi` feature, so probably an issue there.

**Comment by geomatsi** [comment]
> I've hit some weird corner case when doing a large SPI transaction (~100kB in 1kB parts) followed by a pin read. The GPIO read returns no data bytes and panics.
> 
> Inserting a delay of a few milliseconds between the transfer and GPIO read solves the issue.
> 
> I'll try to make an example to reproduce it later today.

I also observed similar issues with `libftdi` during our work on mpsse layer: see pull-request https://github.com/tanriol/ftdi-rs/pull/16. I used flush operation as a workaround. Here is my comment from `ftdi-rs` pull-request:
`
I don't have a good explanation for this particular flush operation at the moment. This flush is not needed when FTDI proprietary libftd2xx library is used. However with libftdi library this flush is needed. Otherwise several first readings from FTDI chip are broken: e.g. the read operation after the first exchange returns nothing, but the read operation after the second exchange returns both values. For the second revision I replaced full flush with Tx flush.
`
I guess a deeper look into `libftdi` is needed to figure out what goes wrong...

**Comment by geomatsi** [comment]
Hi @Atmelfan 

I did a minor cleanup of your PR keeping your authorship. List of changes:
* squashed fixup commits
* fixed doc comments for gpio input methods
* split gpio input into a separate example

Let me know your are ok with the current PR shape.

Regards,
Sergey

**Review by geomatsi — APPROVED** [review]

**Comment by Atmelfan** [comment]
Looks good to me.

**Review by newAM — APPROVED** [review]
Changes look good to me!  I want to check it once more with a physical device before merging; I will be able to do that tomorrow evening.  @geomatsi feel free to merge if you have already checked it with physical hardware.

**Comment by geomatsi** [comment]
> Changes look good to me! I want to check it once more with a physical device before merging; I will be able to do that tomorrow evening. @geomatsi feel free to merge if you have already checked it with physical hardware.

Sure, I have checked ft232/ft2232 devices and the following examples: at24c04, lm75, spi-flash, ws2812, and input.

**Comment by newAM** [comment]
Released [v0.11.0](https://github.com/ftdi-rs/ftdi-embedded-hal/tree/v0.11.0) with these changes.  Thanks for the pull request!

**Pull request merged**